src/Raven.Server/Documents/Archive/ArchivalStorage.cs

```diff
@@ -0,0 +1,341 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Raven.Client;
+using Raven.Client.Documents.Attachments;
+using Raven.Client.Documents.Operations.Counters;
+using Raven.Server.Documents.Replication.ReplicationItems;
+using Raven.Server.Documents.TimeSeries;
+using Raven.Server.ServerWide.Context;
+using Sparrow;
+using Sparrow.Json;
+using Sparrow.Logging;
+using Voron;
+using Sparrow.Json.Parsing;
+using Voron.Impl;
+using Bits = Sparrow.Binary.Bits;
+
+namespace Raven.Server.Documents.Archival
+{
+    public unsafe class ArchivalStorage
+    {
+        private const string DocumentsByArchive = "DocumentsByArchive";
+        
+        private readonly DocumentDatabase _database;
+        private readonly DocumentsStorage _documentsStorage;
+        private readonly Logger _logger;
+
+        public ArchivalStorage(DocumentDatabase database, Transaction tx)
+        {
+            _database = database;
+            _documentsStorage = _database.DocumentsStorage;
+            _logger = LoggingSource.Instance.GetLogger<ArchivalStorage>(database.Name);
+
+            tx.CreateTree(DocumentsByArchive);
+        }
+
+        public void Put(DocumentsOperationContext context, Slice lowerId, BlittableJsonReaderObject document)
+        {
+            if (document.TryGet(Constants.Documents.Metadata.Key, out BlittableJsonReaderObject metadata) == false)
+                return;
+
+            var hasArchiveDate = metadata.TryGet(Constants.Documents.Metadata.Archive, out string archiveDate);
+            if (hasArchiveDate == false)
+                return;
+
+            PutInternal(context, lowerId, archiveDate);
+        }
+
+        private void PutInternal(DocumentsOperationContext context, Slice lowerId, string archiveDate)
+        {
+            if (DateTime.TryParseExact(archiveDate, DefaultFormat.DateTimeFormatsToRead, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out DateTime date) == false) 
+                ThrowWrongArchiveDateFormat(lowerId, archiveDate);
+
+            var archives = date.ToUniversalTime();
+            var ticksBigEndian = Bits.SwapBytes(archives.Ticks);
+
+            var tree = context.Transaction.InnerTransaction.ReadTree(DocumentsByArchive);
+            using (Slice.External(context.Allocator, (byte*)&ticksBigEndian, sizeof(long), out Slice ticksSlice))
+                tree.MultiAdd(ticksSlice, lowerId);
+        }
+
+        private void ThrowWrongArchiveDateFormat(Slice lowerId, string archiveDate)
+        {
+            throw new InvalidOperationException(
+                $"The archive date format for document '{lowerId}' is not valid: '{archiveDate}'. Use the following format: {_database.Time.GetUtcNow():O}");
+        }
+
+        public record ArchivedDocumentsOptions
```

Comment:
Why is it a record? Can it be private?

Reply:
Due to many similarities between archival and expiration, I've made it similar to these expiration options, which are defined as a public record 
https://github.com/ravendb/ravendb/blob/adc7fa2bca66e780d3c31c8b3a3c8e3073c458b2/src/Raven.Server/Documents/Expiration/ExpirationStorage.cs#L79-L84